### PR TITLE
Collapse teaser-thread to a single record when the body is the entire hook

### PR DESCRIPTION
**Pull request by kraftbj**
## Proposed changes

When the teaser-thread strategy's default would publish only `[ entire-body, "Continue reading: <permalink>" ]`, the CTA reply is redundant — there's nothing past the hook to "continue reading" to. Collapse to a single record (body text + link-card embed) so the reader gets one clean post; the link-back is preserved on the same record's embed card.

### Real-world example

Test post at `https://bsky.app/profile/bktest2026.bsky.social/post/3ml4torlpect2`: title + body "this is a test." (15 chars). Currently ships as a 2-post self-reply. After this PR: ships as a single post.

### Design after review iteration

1. **Collapse decision is made on the unfiltered default.** The `atmosphere_teaser_thread_posts` filter only runs when the default would NOT collapse. This avoids silently dropping the second entry for filter authors who edited only the hook (or whose 2-entry shape happens to match the redundant fallback by coincidence).

2. **Backward-compat for existing live threads via a stored-count hint.** `build_long_form_records()` now takes an optional `int $stored_count = 0`. `Publisher::update_post()` passes the stored count so the collapse is skipped when it would force a destructive `rewrite_thread()` that re-mints the root URI and orphans external Bluesky engagement (replies, likes, reposts). New posts still get the optimised single-record shape; live 2-record posts keep their existing shape forever.

### Migration semantics

- **New posts** under `teaser-thread` with short body (no `post_excerpt`, body fits in one record): publish as a single record. No CTA reply.
- **Existing 2-record posts**: keep publishing as 2 records on every subsequent update. No engagement loss. The "redundant CTA reply" stays for those posts indefinitely — fixing it retroactively would require re-minting the root URI, which is the bigger evil.
- **Filter authors** wanting a 2-entry custom shape: works for any post with a usable `post_excerpt` (>= 10 chars) or a body long enough to produce a body chunk. Short-body posts with empty excerpts always collapse before the filter runs.

The filter author behavior (point 1) is intentional and pinned by `test_build_long_form_records_teaser_thread_short_body_collapse_pre_empts_filter`.

The backward-compat behavior (point 2) is pinned by `test_build_long_form_records_teaser_thread_short_body_does_not_collapse_when_stored_count_two`.

### Other information

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions

1. Site running `atmosphere_long_form_composition` returning `'teaser-thread'`. (FOSSE sites do this by default; standalone Atmosphere sites need a `mu-plugin` filter.)
2. Publish a normal post type with a title and a short body (under ~280 chars) and **no** `post_excerpt`. Example: title "Test", body "this is a test."
3. Verify on bsky.app: a single post with the body as text and a link-card pointing back to the WP post — no "Continue reading" reply.
4. Publish a longer post (body > 300 chars) and verify the 3-entry teaser thread (hook + body chunk + CTA) still works as before.
5. Publish a short body WITH a usable `post_excerpt` and verify the 2-entry `[ excerpt, CTA ]` shape (excerpt and body are different content; the CTA still adds value).
6. Backward-compat check: a post that was previously published as a 2-record thread (you'll need to test this against trunk before this PR's changes land, edit it after, verify the URI is preserved). Or, simulate by setting `META_THREAD_RECORDS` to a 2-entry array on a freshly-created post, calling `Publisher::update_post()`, and verifying the in-place update path is taken (1 atomic apply_writes call, no delete-and-republish).

### Changelog

`.github/changelog/teaser-thread-collapse-redundant-cta` (Significance: patch, Type: fixed).

### CODE REVIEW

File: includes/transformer/class-post.php [hunk 1]
```diff
@@ -797,6 +824,39 @@ private function has_composable_body(): bool {
...
+			return false;
+		}
+
+		$excerpt = sanitize_text( (string) $this->object->post_excerpt );
+		if ( \mb_strlen( $excerpt ) >= 10 ) {
+			return false;
+		}
+
+		return ( $texts[1] ?? '' ) === $this->teaser_thread_cta_text();
+	}
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/transformer/class-post.php]: Resolved by the design pivot: collapse decision is now made on the unfiltered default (before `apply_filters('atmosphere_teaser_thread_posts', ...)` runs). A filter that returns `[ hook, default_CTA ]` no longer triggers a silent collapse — the filter doesn't run at all in the collapse case. See `test_build_long_form_records_teaser_thread_short_body_collapse_pre_empts_filter` and `test_build_long_form_records_teaser_thread_filter_runs_when_default_is_not_redundant`.

[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/transformer/class-post.php:943]: Resolved by the design pivot: the predicate no longer compares the second entry's text to the CTA at all. It now checks: count of unfiltered default === 2 AND post_excerpt is empty/<10 chars. Locale, NBSP, and translation drift can't break the check. See `default_teaser_thread_is_redundant_two_entry()`.

File: includes/transformer/class-post.php [hunk 2]
```diff
@@ -460,8 +463,32 @@ public function build_long_form_records(): array {
...
+						$this->record_for_thread_entry(
+							(string) $texts[0],
+							true,
+							array(
+								'strategy'        => 'teaser-thread',
+								'thread_index'    => 0,
+								'is_thread_reply' => false,
+							),
+							$this->build_embed()
+						),
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/transformer/class-post.php]: Fixed in 8f879ff. The docblock now describes both cases: terminal CTA in a multi-record thread, AND the root of a collapsed single-record thread.

File: includes/transformer/class-post.php [hunk 3]
```diff
@@ -797,6 +819,101 @@ private function has_composable_body(): bool {
...
+	 * @return bool
+	 */
+	private function default_teaser_thread_is_redundant_two_entry( array $default_texts ): bool {
+		if ( 2 !== \count( $default_texts ) ) {
+			return false;
+		}
+
+		$excerpt = sanitize_text( (string) $this->object->post_excerpt );
+
+		return \mb_strlen( $excerpt ) < 10;
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/transformer/class-post.php]: Fixed in cc08013. The predicate now also checks `mb_strlen(plain_body) <= 280` (mirroring the body-as-hook budget in `compute_default_teaser_thread()`) so it only fires when the hook IS the whole body, not a truncated prefix. New test `test_build_long_form_records_teaser_thread_truncated_hook_short_tail_does_not_collapse` pins the boundary case (285-char body with 5-char tail → 2-entry default → no collapse).

File: includes/transformer/class-post.php [hunk 4]
```diff
@@ -460,8 +475,48 @@ public function build_long_form_records(): array {
...
+								'strategy'        => 'teaser-thread',
+								'thread_index'    => 0,
+								'is_thread_reply' => false,
+							),
+							$this->build_embed()
+						),
+					);
+				}
+
 				$texts   = $this->build_teaser_thread();
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/transformer/class-post.php]: Fixed in cc08013. `build_teaser_thread()` now takes an optional `?array $precomputed_default = null` parameter. `build_long_form_records()` passes the precomputed default so `render_post_content_plain()` and `truncate_to_budget()` only run once per publish on the non-collapse path.

File: includes/class-publisher.php [hunk 5]
```diff
@@ -559,9 +559,16 @@ public static function update_post( \WP_Post $post ): array|\WP_Error {
...
+		// transformer can preserve the existing thread shape on update
+		// instead of triggering a destructive `rewrite_thread()` for
+		// shape-shrinking optimisations like the redundant-CTA collapse
+		// — that path would re-mint the root URI and orphan external
+		// engagement on the original. New posts (no stored records) get
+		// the optimised shape; live posts keep theirs forever.
 		$new_records = $bsky_transformer->is_short_form_post()
 			? array( $bsky_transformer->transform() )
-			: $bsky_transformer->build_long_form_records();
+			: $bsky_transformer->build_long_form_records( \count( $stored ) );
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/class-publisher.php:590]: Added in 3b50792: `test_update_thread_short_body_in_place_when_stored_two_records`. Seeds `META_THREAD_RECORDS` with 2 records, configures the short-body / empty-excerpt fixture (which would collapse on a fresh publish), and asserts `Publisher::update()` takes the in-place `applyWrites#update` path (not `rewrite_thread`), with TIDs and URIs preserved.

**Review by pfefferle — APPROVED** [review]
Approving. Two non-blocking findings worth noting:

**1. Duplicate `render_post_content_plain()` on the collapse path.**
`compute_default_teaser_thread()` renders the body once at `class-post.php:220`, then `default_teaser_thread_is_redundant_two_entry()` renders it again at `class-post.php:302` for its body-length guard. Not a regression and only on the publish path, but the rendered plain text could be threaded through (or memoised on the instance) so the collapse predicate reuses the already-computed string. Filed as cleanup, not blocking.

**2. `atmosphere_teaser_thread_posts` is silently bypassed when the collapse fires.**
Pinned by `test_build_long_form_records_teaser_thread_short_body_collapse_pre_empts_filter`, so the behaviour is intentional. Flagging it here because filter authors whose callbacks specifically targeted short-body posts will no longer see their hook invoked — the design rationale is well documented inline, but it's the one behaviour change in this PR that a filter author wouldn't infer from the changelog entry. A short release-notes mention (or a `@since` line in the filter docblock pointing at the collapse precondition) would be enough.

Full suite (270 tests / 861 assertions) and `composer lint` green locally on the PR branch.

**Comment by kraftbj** [comment]
Thanks! Handling the first via a follow-up. I'm going to ignore the second finding; the plugin hasn't shipped into the wild yet, so downstream filters are pretty much only internal at this stage or very early adopter.

**Comment by kraftbj** [comment]
Skipping the duplicate-render finding too, but for a different reason — on closer read it doesn't quite match the merged code:

- The functions referenced (`compute_default_teaser_thread()`, `default_teaser_thread_is_redundant_two_entry()`) don't exist. The actual names are `build_teaser_thread()` and `teaser_thread_two_entry_is_redundant()`.
- The collapse predicate (`includes/transformer/class-post.php:842-855`) doesn't call `render_post_content_plain()` — it only checks `count($texts)`, reads `$this->object->post_excerpt`, and string-compares `$texts[1]` to the CTA. So there's no second render on the collapse path.
- The genuine duplicate calls that do exist (`has_composable_body()` at :797 and `build_teaser_thread()` at :689) are already O(1): `Base::render_post_content_plain()` memoizes per-instance via `$plain_content_cache` (`includes/transformer/class-base.php:159-170`), so `the_content` runs once per post per transformer instance regardless of how many helpers ask for the plain text.

My guess is you were looking at an intermediate commit before the predicate was simplified. No follow-up needed on my end — happy to take another look if I'm missing something.
